### DISCUSSION TITLE
Cannot resolve dependency 'http2' with grpc and grpc-transport

### DISCUSSION THREAD

**Discussion by JordanGreissman** [timostamm/protobuf-ts#310] [General]
Trying to install `@protobuf-ts/grpc-transport` in an app that uses create-react-app, I get `cannot resolve dependency 'http2'`. It looks like create-react-app defines an empty mock for http2 in its webpack config, and trying to manually install http2 leads to a critical dependency error: `Critical dependency: require function is used in a way in which dependencies cannot be statically extracted`.

Issues in the grpc-js and create-react-app repos ([here](https://github.com/grpc/grpc-node/issues/610) and [here](https://github.com/facebook/create-react-app/pull/5686)) imply that this has been solved in the version of both that I'm installing, so I'm unsure what to do. Has anyone encountered this?

node version: 14.16.0
npm version: 6.14.11
create-react-app version: 3.4.4

**Comment by timostamm** [comment]
Just in case: gRPC doesn't work in the browser. I assume you want to do server side rendering?

The issue seems to be: `@protobuf-ts/grpc-transport` uses the package `@grpc/grpc-js` as the gRPC implementation. `@grpc/grpc-js` requires the Node.js built-in module `http2`. You cannot install it, you either get it with Node.js, or you don't have it. 

Converting to a discussion, since this is not really an issue with protobuf-ts. Hope a helpful answer turns up.

**Reply by parkwherever** [reply]
Dumb question: can you help me understand what prevents gRPC from working in the browser. Most modern versions of browsers support http2 right? 

Related - if gRPC doesn't work in the browser, given your familiarity here, what is your recommendation to implement bi-directional streaming between the browser and a backend?

**Reply by timostamm** [reply]
Browsers do not expose trailers to JavaScript, but trailers are essential for gRPC, because that's where it puts the status.

It's sad, really. You can actually make fetch requests to a gRPC service, but you would never know if the service reported an error status, so it's impractical. 

Client streams are also not well supported. See [this article](https://web.dev/fetch-upload-streaming/) for details.

The first step in implementing bidi streaming is to ask yourself if you _really_ need it, or if you can actually model the problem with a server streaming method. If you cannot, I recommend to look at WebSockets. They have very good browser support, and you could implement a websocket transport for protobuf-ts. See [here](https://github.com/improbable-eng/grpc-web/blob/master/client/grpc-web/src/transports/websocket/websocket.ts) for how improbable does it.
